src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -1852,6 +1851,110 @@ let BuildFieldMap (cenv: cenv) env isPartial ty (flds: ((Ident list * Ident) * '
                 | _ -> error(Error(FSComp.SR.tcRecordFieldInconsistentTypes(), m)))
     tinst, tcref, fldsmap, List.rev rfldsList
 
+/// Merges updates to nested record fields on the same level in record copy-and-update.
+///
+/// `TransformAstForNestedUpdates` expands `{ x with A.B = 10; A.C = "" }`
+///
+/// into
+///
+/// { x with
+///        A = { x.A with B = 10 };
+///        A = { x.A with C = "" }
+/// }
+///
+/// which we here convert to
+///
+/// { x with A = { x.A with B = 10; C = "" } }
+let GroupUpdatesToNestedFields (fields: ((Ident list * Ident) * SynExpr option) list) =
```

Comment:
I think more these helpers to a separate file `CheckRecordSyntaxHelpers.fs` if they're independent.